standalone_examples/docker_ce.py

```diff
@@ -0,0 +1,83 @@
+from pyinfra import host
+from pyinfra.modules import apt, server, python, init
+
+# Standalone example to show how to install Docker CE using
+# https://docs.docker.com/install/linux/docker-ce/ubuntu/
+
+SUDO = True
+
+if host.fact.linux_name == 'Ubuntu':
+
+    apt.packages(
+        {'Ensure old docker packages are not present'},
+        [
+            'docker',
+            'docker-engine',
+            'docker.io',
+            'containerd runc',
+        ],
+        present=False,
+    )
+
+    apt.packages(
+        {'Ensure Docker CE prerequisites are present'},
+        [
+            'apt-transport-https',
+            'ca-certificates',
+            'curl', 
+            'gnupg-agent',
+            'software-properties-common',
+        ],
+        update=True,
+    )
+
+    docker_key_exists = False
+    stdout = host.fact.command('apt-key list')
+    if 'Docker' in stdout:
+        docker_key_exists = True
+
+    if not docker_key_exists:
+        apt.key(
+            {'Add the Docker apt gpg key if we need to'},
+            key='https://download.docker.com/linux/ubuntu/gpg',
+        )
+
+    apt.repo(
+        {'Add the Docker CE apt repo'},
+        (
+            'deb [arch=amd64] https://download.docker.com/linux/'
+            '{{ host.fact.lsb_release.id|lower }} '
+            '{{ host.fact.lsb_release.codename }} stable'
+        ),
+        filename='docker-ce-stable',
+    )
+
+    apt.packages(
+        {'Ensure Docker CE is installed'},
+        [
+            'docker-ce',
+            'docker-ce-cli',
+            'containerd.io',
+        ],
+        update=True,
+    )
+
+    init.service(
+        {'Ensure docker service is running'},
+        'docker',
+        running=True,
+        enabled=True,
+    )
+
+    # TODO: is NotImplementedEror the right value?
+    # TODO: show the stdout if there is an error
+    # TODO: when you change the docker to dockerf (to make it not found),
+    #       the output is not in sequence; is that output or does it really
+    #       not run these sequentially?
+    stdout = host.fact.command('docker run hello-world')
```

Comment:
Any `host.fact` will be run _before_ anything else in the "fact" phase (https://pyinfra.readthedocs.io/en/latest/deploy_process.html#executing-deploys). A more suitable check could be implemented using `python.call`:

```py
from pyinfra.modules import python
...

def check_docker_works(state, host):
    status, stdout, stderr = host.run_shell_command('docker run hello-world')
    if not status or 'Hello from Docker!' not in stdout:
        raise Exception('`docker run hello-world` did not work as expected')

python.call(check_docker_works, hello='world')
```

Taken from the example here: https://pyinfra.readthedocs.io/en/latest/examples/dynamic_execution_deploy.html

(also used a plain `Exception` which is probably most appropriate here!)

Reply:
Hmm.. I tried something like that but did not work as expected.

Here's a standalone example:

```
from pyinfra.modules import apt, python, server

def some_command(state, host):
    # Run a command on the server and collect status, stderr and stdout
    status, stdout, stderr = host.run_shell_command('echo hello')
    assert status == 0  # ensure the command executed OK
    returned = stdout[0]
    print("returned:", returned)

python.call(
    {'Run some command'},
    some_command,
)
server.shell(
    {'Execute some shell'},
    'echo "back to other operations!"',
)
```

And the error:
```
$ pyinfra 192.168.2.232 --user vagrant --password vagrant mike.py
--> Loading config...
--> Loading inventory...

--> Connecting to hosts...
    [192.168.2.232] Connected

--> Preparing operations...
    Loading: mike.py
    [192.168.2.232]  Ready: mike.py

--> Proposed changes:
    Ungrouped:
    [192.168.2.232]   Operations: 2   Commands: 2

--> Beginning operation run...
--> Starting operation: Run some command
    [192.168.2.232] Unexpected error in Python callback: TypeError("run_shell_command() missing 1 required positional argument: 'command'",)
    [192.168.2.232] Error
--> pyinfra error: No hosts remaining!
```

What am I missing?

Reply:
I had to pass "state" to run_shell_command... weird.

Reply:
Also, status returns a boolean so, need to update https://github.com/Fizzadar/pyinfra/blob/master/docs/examples/dynamic_execution_deploy.md

Reply:
Sorry @mkinney this is my bad, I've now fixed those bugs in the example.